backend/src/schema/resolvers/images/images.js

```diff
@@ -0,0 +1,121 @@
+  // This behaviour differs from `mergeImage`. If you call `mergeImage`
+  // with metadata for an image that does not exist, it's an indicator
+  // of an error (so throw an error). If we bulk delete an image, it
+  // could very well be that there is no image for the resource.
+  if (image) deleteImageFile(image, deleteCallback)
+  return image
+}
+
+export async function mergeImage(resource, relationshipType, imageInput, opts = {}) {
+  if (typeof imageInput === 'undefined') return
+  if (imageInput === null) return deleteImage(resource, relationshipType, opts)
+  sanitizeRelationshipType(relationshipType)
+  const { transaction, uploadCallback, deleteCallback } = opts
+  if (!transaction)
+    return wrapTransaction(mergeImage, [resource, relationshipType, imageInput], opts)
+
+  let txResult
+  txResult = await transaction.run(
+    `
+    MATCH (resource {id: $resource.id})-[:${relationshipType}]->(image:Image)
+    RETURN image {.*}
+    `,
+    { resource },
+  )
+  const [existingImage] = txResult.records.map(record => record.get('image'))
+  const { upload } = imageInput
+  if (!(existingImage || upload)) throw new UserInputError('Cannot find image for given resource')
+  if (existingImage && upload) deleteImageFile(existingImage, deleteCallback)
+  const url = await uploadImageFile(upload, uploadCallback)
+  const { alt, sensitive, aspectRatio } = imageInput
+  const image = { alt, sensitive, aspectRatio, url }
+  txResult = await transaction.run(
+    `
+    MATCH (resource {id: $resource.id})
+    MERGE (resource)-[:${relationshipType}]->(image:Image)
+    ON CREATE SET image.createdAt = toString(datetime())
+    ON MATCH SET image.updatedAt = toString(datetime())
+    SET image += $image
+    RETURN image {.*}
+    `,
+    { resource, image },
+  )
+  const [mergedImage] = txResult.records.map(record => record.get('image'))
+  return mergedImage
+}
+
+const wrapTransaction = async (wrappedCallback, args, opts) => {
+  const session = getDriver().session()
+  try {
+    const result = await session.writeTransaction(async transaction => {
+      return wrappedCallback(...args, { ...opts, transaction })
+    })
+    return result
+  } finally {
+    session.close()
+  }
+}
+
+const deleteImageFile = (image, deleteCallback = localFileDelete) => {
+  const { url } = image
+  deleteCallback(url)
+  return url
+}
+
+const uploadImageFile = async (upload, uploadCallback = localFileUpload) => {
+  if (!upload) return undefined
+  const { createReadStream, filename, mimetype } = await upload
+  const { name, ext } = path.parse(filename)
+  const uniqueFilename = `${uuid()}-${slug(name)}${ext}`
+
+  return uploadCallback({
+    createReadStream,
+    destination: `/uploads/${uniqueFilename}`,
+    mimetype,
+  })
+}
+
+const sanitizeRelationshipType = relationshipType => {
+  // Cypher query language does not allow to parameterize relationship types
+  // See: https://github.com/neo4j/neo4j/issues/340
```

Comment:
> [<img alt="mattwr18" height="40" width="40" align="left" src="https://avatars1.githubusercontent.com/u/38261864?s=88&v=4">](https://github.com/mattwr18) **Authored by [mattwr18](https://github.com/mattwr18)**
_<time datetime="2020-02-25T15:44:57Z" title="Tuesday, February 25th 2020, 4:44:57 pm +01:00">Feb 25, 2020</time>_

---
> **Outdated (history rewrite)** - original diff
---
```diff
@@ -0,0 +1,120 @@
+import path from 'path'
+import uuid from 'uuid/v4'
+import slug from 'slug'
+import { existsSync, unlinkSync, createWriteStream } from 'fs'
+import { getDriver } from '../../../db/neo4j'
+import { UserInputError } from 'apollo-server'
+
+// const widths = [34, 160, 320, 640, 1024]
+
+export async function deleteImage(resource, relationshipType, opts = {}) {
+  sanitizeRelationshipType(relationshipType)
+  const { transaction, deleteCallback } = opts
+  if (!transaction) return wrapTransaction(deleteImage, [resource, relationshipType], opts)
+  const txResult = await transaction.run(
+    `
+    MATCH (resource {id: $resource.id})-[rel:${relationshipType}]->(image:Image)
+    WITH image, image {.*} as imageProps
+    DETACH DELETE image
+    RETURN imageProps
+  `,
+    { resource },
+  )
+  const [image] = txResult.records.map(record => record.get('imageProps'))
+  // This behaviour differs from `mergeImage`. If you call `mergeImage`
+  // with metadata for an image that does not exist, it's an indicator
+  // of an error (so throw an error). If we bulk delete an image, it
+  // could very well be that there is no image for the resource.
+  if (image) deleteImageFile(image, deleteCallback)
+  return image
+}
+
+export async function mergeImage(resource, relationshipType, imageInput, opts = {}) {
+  if (typeof imageInput === 'undefined') return
+  if (imageInput === null) return deleteImage(resource, relationshipType, opts)
+  sanitizeRelationshipType(relationshipType)
+  const { transaction, uploadCallback, deleteCallback } = opts
+  if (!transaction)
+    return wrapTransaction(mergeImage, [resource, relationshipType, imageInput], opts)
+
+  let txResult
+  txResult = await transaction.run(
+    `
+    MATCH (resource {id: $resource.id})-[:${relationshipType}]->(image:Image)
+    RETURN image {.*}
+    `,
+    { resource },
+  )
+  const [existingImage] = txResult.records.map(record => record.get('image'))
+  const { upload } = imageInput
+  if (!(existingImage || upload)) throw new UserInputError('Cannot find image for given resource')
+  if (existingImage && upload) deleteImageFile(existingImage, deleteCallback)
+  const url = await uploadImageFile(upload, uploadCallback)
+  const { alt, sensitive, aspectRatio } = imageInput
+  const image = { alt, sensitive, aspectRatio, url }
+  txResult = await transaction.run(
+    `
+    MATCH (resource {id: $resource.id})
+    MERGE (resource)-[:${relationshipType}]->(image:Image)
+    SET image += $image
+    SET image.createdAt = toString(datetime())
+    RETURN image {.*}
+    `,
+    { resource, image },
+  )
+  const [mergedImage] = txResult.records.map(record => record.get('image'))
+  return mergedImage
+}
+
+const wrapTransaction = async (wrappedCallback, args, opts) => {
+  const session = getDriver().session()
+  try {
+    const result = await session.writeTransaction(async transaction => {
+      return wrappedCallback(...args, { ...opts, transaction })
+    })
+    return result
+  } finally {
+    session.close()
+  }
+}
+
+const deleteImageFile = (image, deleteCallback = localFileDelete) => {
+  const { url } = image
+  deleteCallback(url)
+  return url
+}
+
+const uploadImageFile = async (upload, uploadCallback = localFileUpload) => {
+  if (!upload) return undefined
+  const { createReadStream, filename, mimetype } = await upload
+  const { name, ext } = path.parse(filename)
+  const uniqueFilename = `${uuid()}-${slug(name)}${ext}`
+
+  return uploadCallback({
+    createReadStream,
+    destination: `/uploads/${uniqueFilename}`,
+    mimetype,
+  })
+}
+
+const sanitizeRelationshipType = relationshipType => {
+  // Cypher query language does not allow to parameterize relationship types
+  // See: https://github.com/neo4j/neo4j/issues/340
```


I looked at this issue, thanks for linking to it! this is one case I like a comment here.
can't we just use `MERGE (resource)-[:HERO_IMAGE|AVATAR_IMAGE]->(image:Image)`
See https://neo4j.com/docs/cypher-manual/current/clauses/match/#match-on-multiple-rel-types

Reply:
> [<img alt="roschaefer" height="40" width="40" align="left" src="https://avatars1.githubusercontent.com/u/38261864?s=88&v=4">](https://github.com/roschaefer) **Authored by [roschaefer](https://github.com/roschaefer)**
_<time datetime="2020-03-02T16:27:26Z" title="Monday, March 2nd 2020, 5:27:26 pm +01:00">Mar 2, 2020</time>_

---


that's `MATCH` not `MERGE`

Reply:
> [<img alt="mattwr18" height="40" width="40" align="left" src="https://avatars1.githubusercontent.com/u/38261864?s=88&v=4">](https://github.com/mattwr18) **Authored by [mattwr18](https://github.com/mattwr18)**
_<time datetime="2020-03-11T19:25:49Z" title="Wednesday, March 11th 2020, 8:25:49 pm +01:00">Mar 11, 2020</time>_

---


yep, I remember we discussed this